applied-crypto-examples/src/encrypted_zksnark.rs

```diff
@@ -0,0 +1,254 @@
+            }
+            //println!("coefficients: {:?}", coefficients);
+        }
+        coefficients.reverse();
+        Ok(Self {
+            roots,
+            coefficients,
+            num_public_roots,
+        })
+    }
+
+    /// Get degree of polynomial
+    pub fn degree(&self) -> usize {
+        self.roots.len()
+    }
+
+    /// calculate encrypted polynomial
+    pub fn generate_response(&self, powers: &EncryptedChallenge) -> EncryptedResponse {
+        let px = self.eval(powers);
+        let h = Polynomial::new(self.roots[self.num_public_roots..].to_vec(), 1).unwrap();
+        let hx = h.eval(powers);
+        EncryptedResponse::new(px, hx)
+    }
+
+    /// Evaluate polynomial at given encrypted powers
+    pub fn eval(&self, powers: &EncryptedChallenge) -> RistrettoPoint {
+        powers
+            .encrypted_powers
+            .iter()
+            .zip(self.coefficients.iter())
+            .map(|(p, c)| p * c)
+            .sum()
+    }
+
+    /// Create public polynomial from private polynomial
+    pub fn eval_public_polynomial(&self, scalar: &Scalar) -> Scalar {
+        self.roots[0..self.num_public_roots]
+            .to_vec()
+            .iter()
+            .fold(Scalar::one(), |acc, root| acc * root.eval(scalar))
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_encrypted_polynomial_roots_must_divide() {
+        assert_eq!(Root::new(2, 1), Err(Error::OutsideIntegerField(2, 1)));
+    }
+
+    #[test]
+    fn test_encrypted_powers_calculate_correctly() {
+        let roots = vec![
+            Root::new(1, 2).unwrap(),
+            Root::new(3, 6).unwrap(),
+            Root::new(2, 4).unwrap(),
+            Root::new(1, 6).unwrap(),
+            Root::new(2, 6).unwrap(),
+        ];
+
+        let polynomial = Polynomial::new(roots, 2).unwrap();
+        let scalar = Scalar::from(5u64);
+        let powers = EncryptedChallenge::new(&scalar, &polynomial);
+        assert_eq!(powers.encrypted_powers.len(), 6);
+        assert_eq!(powers.encrypted_powers[0], G);
+        assert_eq!(powers.encrypted_powers[1], Scalar::from(5u64) * G);
+        assert_eq!(powers.encrypted_powers[2], Scalar::from(25u64) * G);
+        assert_eq!(powers.encrypted_powers[3], Scalar::from(125u64) * G);
+        assert_eq!(powers.encrypted_powers[4], Scalar::from(625u64) * G);
+        assert_eq!(powers.encrypted_powers[5], Scalar::from(3125u64) * G);
+    }
+
+    #[test]
+    fn test_encrypted_polynomial_evaluates_correctly() {
+        let roots = vec![
+            Root::new(1, 2).unwrap(),
+            Root::new(3, 6).unwrap(),
+            Root::new(2, 4).unwrap(),
+        ];
+
+        let polynomial = Polynomial::new(roots, 2).unwrap();
+        let scalar = Scalar::from(5u64);
+        let challenge = EncryptedChallenge::new(&scalar, &polynomial);
+
+        assert_eq!(polynomial.degree(), 3);
+        assert_eq!(polynomial.eval(&challenge), Scalar::from(2058u64) * G);
+    }
```

Comment:
This should be done without using a public eval.

Reply:
Agreed!